libgeopmd/src/StatsCollector.cpp

```diff
@@ -0,0 +1,194 @@
+                                 struct geopm_stats_collector_s **collector)
+{
+    int err = 0;
+    try {
+        std::vector<geopm_request_s> request_vec(requests, requests + num_requests);
+        auto result = std::make_unique<geopm::StatsCollector>(request_vec);
+        *collector = reinterpret_cast<geopm_stats_collector_s *>(result.release());
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+int geopm_stats_collector_update(struct geopm_stats_collector_s *collector)
+{
+    int err = 0;
+    try {
+        geopm::StatsCollector *collector_cpp = reinterpret_cast<geopm::StatsCollector *>(collector);
+        collector_cpp->update();
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+// If *max_report_size is zero, update it with the required size for the report
+// and do not modify report
+int geopm_stats_collector_report_yaml(const struct geopm_stats_collector_s *collector,
+                                      size_t *max_report_size, char *report_yaml)
+{
+    int err = 0;
+    try {
+        const geopm::StatsCollector *collector_cpp = reinterpret_cast<const geopm::StatsCollector *>(collector);
+        std::string report_str = collector_cpp->report_yaml();
+        if (report_str.size() < *max_report_size) {
+            strncpy(report_yaml, report_str.c_str(), report_str.size());
+        }
+        else {
+            err = ENOBUFS;
+            std::ostringstream err_str;
+            size_t report_str_size = report_str.size() + 1;
+            if (*max_report_size != 0) {
+                err_str << "geopm_stats_collector_report_yaml(): max_report_size is too small, provided: "
+                        << *max_report_size << " required: " << report_str_size;
+            }
+            else {
+                err_str << "geopm_stats_collector_report_yaml(): max_report_size required: " << report_str_size;
+            }
+            *max_report_size = report_str_size;
+            throw geopm::Exception(err_str.str(), err, __FILE__, __LINE__);
+        }
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+int geopm_stats_collector_reset(struct geopm_stats_collector_s *collector)
+{
+    int err = 0;
+    try {
+        geopm::StatsCollector *collector_cpp = reinterpret_cast<geopm::StatsCollector *>(collector);
+        collector_cpp->reset();
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+int geopm_stats_collector_free(geopm_stats_collector_s *collector)
```

Comment:
Minor nit for consistency:

```suggestion
int geopm_stats_collector_free(struct geopm_stats_collector_s *collector)
```

In all other places the `struct` keyword is included.